File: tx3/examples/typescript/submit_swap.ts
import { Client, type Profile, type SubmitSwapParams } from "../../gen/typescript/sundae-v3/protocol";
import { Party } from "tx3-sdk";
import type { ClientOptions } from "tx3-sdk/trp";

const profile: Profile = "preview";

const options: ClientOptions = {
  endpoint: "https://cardano-preview.trp-m1.demeter.run",
  headers: {
    "dmtr-api-key": process.env.DEMETER_TRP_API_KEY ?? "",
  },
};

const userAddress = process.env.SUNDAE_USER_ADDRESS ?? "<wallet address>";

const params: SubmitSwapParams = {
  poolIdent: "35a34996f515c5a28c8df9eada81f03f4f2756d92e7f73cde1f4e593",
  ownerKeyHash: "<stake key hash>",
  destinationPaymentKeyHash: "<payment key hash>",
  destinationStakeKeyHash: "<stake key hash>",
  orderAda: 3_000_000,
  maxProtocolFee: 600_000,
  offerPolicy: "d8906ca5c7ba124a0407a32dab37b2c82b13b3dcd9111e42940dcea4",
  offerName: "0014df105553444d",
  offerAmount: 1_000_000,
  minReceivedPolicy: "",
  minReceivedName: "",
  minReceivedAmount: 1,
};

async function main() {
  if (!options.headers?.["dmtr-api-key"]) {
    throw new Error("missing DEMETER_TRP_API_KEY");
  }

  if (!process.env.SUNDAE_USER_ADDRESS) {
    throw new Error("missing SUNDAE_USER_ADDRESS");
  }

  const client = new Client(options, profile).withUser(Party.address(userAddress));

  // The preview/mainnet profiles already carry the `orderscript` party value.
  // Call `.withOrderscript(...)` only if you want to override it.

  const builder = client.submitSwap(params);
  const resolved = await builder.resolve();

  console.log("resolve ok");
  console.log("tx hash:", resolved.hash);
  console.log("tx hex:", resolved.txHex);
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: tx3/examples/python/submit_swap.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import asyncio

from common import (  # noqa: E402
    DEFAULT_POOL_IDENT,
    DEFAULT_USDM,
    add_common_preview_args,
    make_preview_client,
    parse_asset,
    payment_and_stake_key_hashes,
    require_api_key_and_user,
    resolve_or_submit,
)
from sundae import SubmitSwapParams  # noqa: E402


async def main() -> int:
    parser = argparse.ArgumentParser(description="Resolve or submit a Sundae swap order.")
    add_common_preview_args(parser)
    parser.add_argument("--pool-ident", default=DEFAULT_POOL_IDENT)
    parser.add_argument("--offer", default=DEFAULT_USDM)
    parser.add_argument("--offer-amount", type=int, default=1_000_000)
    parser.add_argument("--receive", default="ada")
    parser.add_argument("--min-received", type=int, default=1)
    parser.add_argument("--order-ada", type=int, default=3_000_000)
    parser.add_argument("--max-protocol-fee", type=int, default=600_000)
    parser.add_argument("--owner-key-hash", default="")
    parser.add_argument("--destination-payment-key-hash", default="")
    parser.add_argument("--destination-stake-key-hash", default="")
    args = parser.parse_args()

    require_api_key_and_user(args)
    client, can_sign = make_preview_client(args)

    offer_policy, offer_name = parse_asset(args.offer)
    receive_policy, receive_name = parse_asset(args.receive)
    payment_kh, stake_kh = payment_and_stake_key_hashes(args.user)

    builder = client.submit_swap(
        SubmitSwapParams(
            pool_ident=args.pool_ident,
            owner_key_hash=args.owner_key_hash or stake_kh,
            destination_payment_key_hash=args.destination_payment_key_hash or payment_kh,
            destination_stake_key_hash=args.destination_stake_key_hash or stake_kh,
            order_ada=args.order_ada,
            max_protocol_fee=args.max_protocol_fee,
            offer_policy=offer_policy,
            offer_name=offer_name,
            offer_amount=args.offer_amount,
            min_received_policy=receive_policy,
            min_received_name=receive_name,
            min_received_amount=args.min_received,
        )
    )

    await resolve_or_submit(builder, args.submit, can_sign)
    return 0


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))
